backend/src/services/mirroredModel.ts

```diff
@@ -0,0 +1,224 @@
+import { createHash } from 'node:crypto'
+
+import archiver from 'archiver'
+import stream, { PassThrough } from 'stream'
+
+import { sign } from '../clients/kms.js'
+import { putObjectStream } from '../clients/s3.js'
+import { ModelAction } from '../connectors/authorisation/actions.js'
+import authorisation from '../connectors/authorisation/index.js'
+import { FileInterfaceDoc, ScanState } from '../models/File.js'
+import { ModelDoc } from '../models/Model.js'
+import { ReleaseDoc } from '../models/Release.js'
+import { UserInterface } from '../models/User.js'
+import config from '../utils/config.js'
+import { BadReq, Forbidden, InternalError } from '../utils/error.js'
+import { downloadFile, getFilesByIds, getTotalFileSize } from './file.js'
+import log from './log.js'
+import { getModelById, getModelCardRevisions } from './model.js'
+import { getAllFileIds, getReleasesForExport } from './release.js'
+
+const ExportKind = {
+  ModelCards: 'modelCards',
+  Releases: 'releases',
+} as const
+type ExportKindKeys = (typeof ExportKind)[keyof typeof ExportKind]
+
+type ZipFile = {
+  filename: string
+  file: archiver.Archiver
+  metadata: { modelId: string; exportKind: ExportKindKeys }
+}
+
+export async function exportModel(
+  user: UserInterface,
+  modelId: string,
+  disclaimerAgreement: boolean,
+  releaseSemvers?: Array<string>,
+) {
+  if (!config.modelMirror.enabled) {
+    throw BadReq('Model mirroring has not been enabled.')
+  }
+  if (!disclaimerAgreement) {
+    throw BadReq('You must agree to the disclaimer agreement before being able to export a model.')
+  }
+  const model = await getModelById(user, modelId)
+  if (!model.settings.mirroredModelId || model.settings.mirroredModelId === '') {
+    throw BadReq('The ID of the mirrored model has not been set on this model.')
+  }
+  const auth = await authorisation.model(user, model, ModelAction.Update)
+  if (!auth.success) {
+    throw Forbidden(auth.info, { userDn: user.dn, model: model.id })
+  }
+
+  const zipFiles: ZipFile[] = []
+  zipFiles.push({
+    filename: `${modelId}/modelCards.zip`,
+    file: await generateModelCardRevisionsZip(user, model),
+    metadata: { modelId, exportKind: ExportKind.ModelCards },
+  })
+  // unknown release semver
+  if (releaseSemvers && releaseSemvers.length > 0) {
+    zipFiles.push({
+      filename: `${modelId}/releases.zip`,
+      file: await generateReleaseZip(user, model, releaseSemvers),
+      metadata: { modelId, exportKind: ExportKind.Releases },
+    })
+  }
+
+  return Promise.all(zipFiles.map((zipFile) => uploadZipFileToS3(zipFile)))
+}
+
+async function uploadZipFileToS3(zip: ZipFile) {
+  const s3Stream = new PassThrough()
+  const hashStream = new PassThrough()
+
+  zip.file.pipe(hashStream)
+  zip.file.pipe(s3Stream)
+
+  let signatures = {}
+  if (config.modelMirror.export.kmsSignature.enabled) {
+    try {
+      const messageDigest = await generateDigest(hashStream)
```

Comment:
I think this will break.

You create a 'zip file' of the zip files and model cards.  Then you wait for it to _entirely_ create before generating a hash.  We won't have enough memory to reliably do this.  Also, during this time the 's3Stream' hasn't done anything, so if your file is larger than the HighWatermark the upload will stop?